crates/worker/src/partition/leadership/trim_queue.rs

```diff
@@ -0,0 +1,496 @@
+// Copyright (c) 2023 - 2025 Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+// todo: remove when trimming is fully implemented in PP
+#![allow(dead_code)]
+
+use std::collections::BTreeMap;
+use std::sync::Arc;
+use std::time::Duration;
+
+use jiff::Timestamp;
+use parking_lot::Mutex;
+use tokio::time::MissedTickBehavior;
+use tracing::{debug, instrument, warn};
+
+use restate_bifrost::Bifrost;
+use restate_core::{ShutdownError, TaskCenter, TaskId, cancellation_token};
+use restate_storage_api::fsm_table::PartitionDurability;
+use restate_types::config::Configuration;
+use restate_types::logs::{LogId, Lsn, SequenceNumber};
+use restate_types::time::MillisSinceEpoch;
+
+/// A task that trims the log by removing durable LSNs from the log.
+pub struct LogTrimmer {
+    bifrost: Bifrost,
+    log_id: LogId,
+    trim_queue: TrimQueue,
+}
+impl LogTrimmer {
+    pub fn spawn(
+        bifrost: Bifrost,
+        log_id: LogId,
+        trim_queue: TrimQueue,
+    ) -> Result<TaskId, ShutdownError> {
+        let task = Self {
+            trim_queue,
+            log_id,
+            bifrost,
+        };
+
+        TaskCenter::spawn_child(
+            restate_core::TaskKind::LogTrimmer,
+            "log-trimmer",
+            task.run(),
+        )
+    }
+
+    async fn run(self) -> anyhow::Result<()> {
+        let cancel = cancellation_token();
+        let mut interval = tokio::time::interval(Duration::from_secs(1));
+        interval.set_missed_tick_behavior(MissedTickBehavior::Delay);
+
+        debug!(log_id = %self.log_id, "Log trimmer task started");
+        loop {
+            if let Some(durability_point) = self.trim_queue.pop() {
+                let Some(result) = cancel
+                    .run_until_cancelled(self.trim_logs(&durability_point))
+                    .await
+                else {
+                    // if cancelled, we push it back. it's safe to do because trims are
+                    // idemponent.
+                    self.trim_queue.push(&durability_point);
+                    break;
+                };
+
+                if !result {
+                    self.trim_queue.push(&durability_point);
+                    // the next retry happens after the sleep.
+                }
```

Comment:
Should we update the `known_trim_point` via `self.trim_queue.notify_trimmed(durability_point.durable_point)` if the trim was successful?